From 7714396460ce00dfcf8315d181db340bd11da0dd Mon Sep 17 00:00:00 2001
From: Teemu Toivola <git@humdi.net>
Date: Sat, 30 Jan 2021 16:36:16 +0200
Subject: [PATCH] add --initdb to vnstatd for creating a new empty database
 without having the daemon process staying running, doesn't discard data if a
 database already exists

---
 src/daemon.c  |  1 +
 src/daemon.h  |  2 +-
 src/vnstatd.c | 19 ++++++++++++++++++-
 5 files changed, 36 insertions(+), 7 deletions(-)

diff --git a/src/daemon.c b/src/daemon.c
index f693111..647efe5 100644
--- a/src/daemon.c
+++ b/src/daemon.c
@@ -253,6 +253,7 @@ void initdstate(DSTATE *s)
 	s->forcesave = 0;
 	s->noadd = 0;
 	s->alwaysadd = 0;
+	s->initdb = 0;
 	s->iflisthash = 0;
 	s->cfgfile[0] = '\0';
 	s->user[0] = '\0';
diff --git a/src/daemon.h b/src/daemon.h
index 44efb47..08b3714 100644
--- a/src/daemon.h
+++ b/src/daemon.h
@@ -4,7 +4,7 @@
 typedef struct {
 	int updateinterval, saveinterval;
 	short running, dodbsave, rundaemon;
-	short dbsaved, showhelp, sync, forcesave, noadd;
+	short dbsaved, showhelp, sync, forcesave, noadd, initdb;
 	short alwaysadd, bootdetected, cleanuphour, dbretrycount;
 	uint32_t iflisthash;
 	uint64_t dbifcount;
diff --git a/src/vnstatd.c b/src/vnstatd.c
index b68d85d..4d46353 100644
--- a/src/vnstatd.c
+++ b/src/vnstatd.c
@@ -83,6 +83,14 @@ int main(int argc, char *argv[])
 		exit(EXIT_FAILURE);
 	}
 
+	if (s.initdb) {
+		db_close();
+		if (debug) {
+			printf("--initdb complete, exiting...\n");
+		}
+		exit(EXIT_SUCCESS);
+	}
+
 	detectboot(&s);
 	preparedatabase(&s);
 
@@ -238,7 +246,8 @@ void showhelp(void)
 	printf("      -g, --group <group>      set daemon process group\n");
 	printf("      --config <config file>   select used config file\n");
 	printf("      --noadd                  don't add found interfaces if no dbs are found\n");
-	printf("      --alwaysadd              always add new interfaces even when some dbs exist\n\n");
+	printf("      --alwaysadd              automatically start monitoring all new interfaces\n");
+	printf("      --initdb                 create empty database and exit\n\n");
 
 	printf("See also \"man vnstatd\".\n");
 }
@@ -289,6 +298,9 @@ void parseargs(DSTATE *s, int argc, char **argv)
 			s->noadd = 1;
 		} else if (strcmp(argv[currentarg], "--alwaysadd") == 0) {
 			s->alwaysadd = 1;
+		} else if (strcmp(argv[currentarg], "--initdb") == 0) {
+			s->initdb = 1;
+			s->showhelp = 0;
 		} else if ((strcmp(argv[currentarg], "-v") == 0) || (strcmp(argv[currentarg], "--version") == 0)) {
 			printf("vnStat daemon %s by Teemu Toivola <tst at iki dot fi>\n", getversion());
 			exit(EXIT_SUCCESS);
@@ -319,6 +331,11 @@ void parseargs(DSTATE *s, int argc, char **argv)
 		exit(EXIT_FAILURE);
 	}
 
+	if (s->rundaemon && s->initdb) {
+		printf("Error: --daemon and --initdb can't both be used at the same time.\n");
+		exit(EXIT_FAILURE);
+	}
+
 	/* show help if nothing else was asked to be done */
 	if (s->showhelp) {
 		showhelp();
